File: src/file_tools/read_file_head.py
#!/usr/bin/env python3
"""Read the first N lines of a file with line numbers. Max lines: 200."""

import os

from _common import emit_error, emit_json, load_args, resolve_path

DEFAULT_LINES = 80
MAX_LINES = 200


def main():
    args = load_args()
    file_path = resolve_path(args.get('path'), args.get('workspace'))
    max_lines = int(args.get('max_lines', DEFAULT_LINES))

    if max_lines < 1:
        emit_error('max_lines must be >= 1')
    if max_lines > MAX_LINES:
        emit_error(f'Refused: max_lines {max_lines} exceeds cap {MAX_LINES}')

    if not os.path.isfile(file_path):
        emit_error(f'File not found: {file_path}')

    lines_out = []
    total_lines = 0
    try:
        with open(file_path, 'r', encoding='utf-8', errors='replace') as handle:
            for line_no, line in enumerate(handle, start=1):
                total_lines = line_no
                if line_no > max_lines:
                    break
                lines_out.append(f'{line_no:6}|{line.rstrip()}')
    except OSError as exc:
        emit_error(f'Failed to read file: {exc}')

    truncated = total_lines > max_lines
    emit_json({
        'path': file_path,
        'max_lines': max_lines,
        'line_count': len(lines_out),
        'total_lines': total_lines if not truncated else f'>{max_lines}',
        'truncated': truncated,
        'text': '\n'.join(lines_out)
    })


if __name__ == '__main__':
    try:
        main()
    except Exception as exc:  # noqa: BLE001
        emit_error(str(exc))

File: src/file_tools/read_file_head.js
/**
 * Read the first N lines of a file with line numbers. Max lines: 200.
 * Pure Node.js in-process implementation.
 */

const fs = require('fs');
const common = require('./common');

const DEFAULT_LINES = 80;
const MAX_LINES = 200;

function readFileHead(args = {}) {
  const rawPath = args.path || args.file_path || args.filepath || args.FilePath || args.Path;
  const filePath = common.resolvePath(rawPath, args.workspace, args.dev_root);
  const maxLines = parseInt(args.max_lines || args.maxLines || DEFAULT_LINES, 10);

  if (isNaN(maxLines) || maxLines < 1) {
    throw new Error('max_lines must be >= 1');
  }
  if (maxLines > MAX_LINES) {
    throw new Error(`Refused: max_lines ${maxLines} exceeds cap ${MAX_LINES}`);
  }

  if (!fs.existsSync(filePath) || !fs.statSync(filePath).isFile()) {
    throw new Error(`File not found: ${filePath}`);
  }

  const content = fs.readFileSync(filePath, 'utf8');
  const allLines = content.split(/\r?\n/);
  const totalLines = allLines.length;

  const MAX_LINE_CHARS = 4000;
  const slice = allLines.slice(0, maxLines);
  const linesOut = slice.map((line, idx) => {
    const lineNo = idx + 1;
    const safeLine = line.length > MAX_LINE_CHARS ? line.substring(0, MAX_LINE_CHARS) + `... [line truncated from ${line.length} chars]` : line;
    return `${String(lineNo).padStart(6, ' ')}|${safeLine}`;
  });

  const truncated = totalLines > maxLines;
  return {
    path: filePath,
    max_lines: maxLines,
    line_count: linesOut.length,
    total_lines: truncated ? `>${maxLines}` : totalLines,
    truncated,
    text: linesOut.join('\n')
  };
}

module.exports = {
  readFileHead,
  run: readFileHead
};
